### PR TITLE
[radio] add coex metrics

### PR DESCRIPTION
**Pull request by zhanglongxia**
1. Add platform get coex metrics API
2. Add spinel property for coex metrics

### CODE REVIEW

File: .travis/check-posix-app-pty [hunk 1]
```diff
@@ -96,6 +96,8 @@ send "extaddr\r\n"
 expect "Done"
 send "ipaddr\r\n"
 expect "Done"
+send "coex\r\n"
```
Comment (bukepo) [inline .travis/check-posix-app-pty]: check the expected output?

File: src/ncp/spinel.h [hunk 2]
```diff
@@ -1054,6 +1054,7 @@ enum
     SPINEL_CAP_CHILD_SUPERVISION       = (SPINEL_CAP_OPENTHREAD__BEGIN + 8),
     SPINEL_CAP_POSIX_APP               = (SPINEL_CAP_OPENTHREAD__BEGIN + 9),
     SPINEL_CAP_SLAAC                   = (SPINEL_CAP_OPENTHREAD__BEGIN + 10),
+    SPINEL_CAP_COEX_METRICS            = (SPINEL_CAP_OPENTHREAD__BEGIN + 11),
```
Comment (bukepo) [inline src/ncp/spinel.h]: I don't think it's necessary to add a new capability. If this features is not enabled, getting this property will responds not found error, which is already enough to indicate this features is not supported. This way, we save code size.

File: src/posix/platform/radio_spinel.cpp [hunk 3]
```diff
@@ -892,6 +892,49 @@ int8_t RadioSpinel::GetRssi(void)
     return rssi;
 }
 
+otError RadioSpinel::GetCoexMetrics(otRadioCoexMetrics *aCoexMetrics)
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: use reference instead of pointer, such that we save line 899.

File: src/cli/cli.cpp [hunk 4]
```diff
@@ -781,6 +784,42 @@ void Interpreter::ProcessCoapSecure(int argc, char *argv[])
...
+    OT_UNUSED_VARIABLE(argc);
+    OT_UNUSED_VARIABLE(argv);
+
+    otRadioCoexMetrics metrics;
+    otError            error = otPlatRadioGetCoexMetrics(mInstance, &metrics);
+
+    SuccessOrExit(error);
+
+    mServer->OutputFormat("Stopped: %s\r\n", metrics.mStopped ? "true" : "false");
+    mServer->OutputFormat("NumGrantGlitch: %d\r\n", metrics.mNumGrantGlitch);
```
Comment (bukepo) [inline src/cli/cli.cpp]: `d` to `u`.

Comment (smides-nest) [reply src/cli/cli.cpp]: If these are not to be scraped (hopefully), then a more readable format is preferred. E.g.:

```
Saturated: false
Grant glitches: xxx
Transmit metrics
    Requests: xxx
    Immediate grants: xxx
    Grant waits: xxx
    Grant waits activated: xxx
    Grant wait timeouts: xxx
    Grant deactivations during request: xxx
    .
    .
    .
Receive metrics
    Requests: xxx
     .
     .
     .
```

Comment (smides-nest) [reply src/cli/cli.cpp]: Actually, I just realized the symmetry of the Tx and Rx stats, so a better format would be

```
                     Tx                    Rx
Requests:            xxx                   yyy
Immediate grants:    xxx                   yyy
.
.
.
```

File: .travis/check-posix-app-pty (hunk 1 above)
Comment (zhanglongxia) [reply .travis/check-posix-app-pty]: I have checked if the output is "Done" to make sure that the function "GetCoexMetrics()" is called.

File: src/ncp/spinel.h (hunk 2 above)
Comment (zhanglongxia) [reply src/ncp/spinel.h]: Great. I have removed the capability property.

File: src/posix/platform/radio_spinel.cpp (hunk 3 above)
Comment (zhanglongxia) [reply src/posix/platform/radio_spinel.cpp]: Updated

File: src/cli/cli.cpp (hunk 4 above)
Comment (zhanglongxia) [reply src/cli/cli.cpp]: @smides-nest The Tx and Rx stats is asymmetrical. The Tx stats doesn't contain `NumTxGrantNone`. So it is better to follow your first suggestion.

**Comment by zhanglongxia** [comment]
@abtink Could you help me review this PR at your convenience?

File: include/openthread/platform/radio.h [hunk 5]
```diff
@@ -706,13 +732,24 @@ uint32_t otPlatRadioGetSupportedChannelMask(otInstance *aInstance);
 /**
  * Get the radio preferred channel mask that the device prefers to form on.
  *
- * @param[in]  aInstance   The OpenThread instance strucyyture.
+ * @param[in]  aInstance   The OpenThread instance structure.
```
Comment (abtink) [inline include/openthread/platform/radio.h]: :+1:

File: src/ncp/spinel.h [hunk 6]
```diff
@@ -1627,6 +1627,43 @@ typedef enum
      */
     SPINEL_PROP_RADIO_CAPS = SPINEL_PROP_PHY_EXT__BEGIN + 11,
 
+    /// All coex metrics related counters.
+    /** Format: bLt(LLLLLLLL)t(LLLLLLLLL)  (Read-only)
+     *
+     * The contents include three variables and two structs, first one corresponds to
```
Comment (abtink) [inline src/ncp/spinel.h]: nit: two variables and two structs?

Comment (abtink) [reply src/ncp/spinel.h]: Suggestion: How about adding the common parameters after the two structs? This way we can later add new fields to the common part while keeping backward compatibility... 

Thoughts?

Comment (zhanglongxia) [reply src/ncp/spinel.h]: Great suggestion. I will update the code.

File: include/openthread/platform/radio.h [hunk 7]
```diff
@@ -706,13 +732,24 @@ uint32_t otPlatRadioGetSupportedChannelMask(otInstance *aInstance);
...
  *
  * @returns The radio preferred channel mask.
  *
  */
 uint32_t otPlatRadioGetPreferredChannelMask(otInstance *aInstance);
 
+/**
+ * Get the radio coexistence metrics.
+ *
+ * @param[in]  aInstance     The OpenThread instance strucyyture.
```
Comment (bukepo) [inline include/openthread/platform/radio.h]: typo: strucyyture

File: src/posix/platform/radio_spinel.cpp [hunk 8]
```diff
@@ -892,6 +892,46 @@ int8_t RadioSpinel::GetRssi(void)
     return rssi;
 }
 
+otError RadioSpinel::GetCoexMetrics(otRadioCoexMetrics &aCoexMetrics)
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: Guard with `#if OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_METRICS_ENABLE` to cause build time error when not properly configured.

File: src/posix/platform/radio_spinel.cpp [hunk 9]
```diff
@@ -1630,6 +1670,20 @@ int8_t otPlatRadioGetReceiveSensitivity(otInstance *aInstance)
     return sRadioSpinel.GetReceiveSensitivity();
 }
 
+otError otPlatRadioGetCoexMetrics(otInstance *aInstance, otRadioCoexMetrics *aCoexMetrics)
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: Guard with #if OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_METRICS_ENABLE to cause build time error when not properly configured.

File: examples/platforms/posix/radio.c [hunk 10]
```diff
@@ -1087,3 +1087,36 @@ int8_t otPlatRadioGetReceiveSensitivity(otInstance *aInstance)
 
     return POSIX_RECEIVE_SENSITIVITY;
 }
+
+otError otPlatRadioGetCoexMetrics(otInstance *aInstance, otRadioCoexMetrics *aCoexMetrics)
```
Comment (bukepo) [inline examples/platforms/posix/radio.c]: Guard with #if OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_METRICS_ENABLE to cause build time error when not properly configured.

Comment (zhanglongxia) [reply examples/platforms/posix/radio.c]: Updated

File: src/posix/platform/radio_spinel.cpp (hunk 8 above)
Comment (zhanglongxia) [reply src/posix/platform/radio_spinel.cpp]: Updated

File: src/posix/platform/radio_spinel.cpp (hunk 9 above)
Comment (zhanglongxia) [reply src/posix/platform/radio_spinel.cpp]: Updated

File: include/openthread/platform/radio.h (hunk 7 above)
Comment (zhanglongxia) [reply include/openthread/platform/radio.h]: Updated.

**Pull request merged**